### PR TITLE
[AutoDiff] Defines remaining derivatives for tgmath functions.

### PR DESCRIPTION
**Pull request by vguerra**
The following math functions are now differentiable:

* `remainder`
* `fmod`
* `ceil`
* `floor`
* `round`
* `trunc`

As well, this PR makes usage of @differentiating instead of
@differentiable attribute for derivate registration.

NOTE: For the time being this exposes a compiler crash that might ( or not )
be related to [TF-429](https://bugs.swift.org/browse/TF-429).

Resolves [TF-812](https://bugs.swift.org/browse/TF-812)

### CODE REVIEW

**Comment by vguerra** [comment]
For some context on this PR, please refer to https://github.com/apple/swift/pull/27953

**Comment by vguerra** [comment]
hi @dan-zheng, could you please trigger the CI. I get locally a compiler crash on this PR :'(

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
I believe stdlib compilation is fixed in https://github.com/apple/swift/pull/28108/commits/f50beec225ac9ef0074ea5ffcc7c393f77a4f4dc!

---

Explanation: currently, functions marked with `@differentiating` must return a tuple with one of the following label schemes:
* `(value: ..., differential: ...)`
* `(value: ..., pullback: ...)`

The `differential:` label indicates that the `@differentiating` function is a forward-mode derivative function, or JVP. It returns a differential function, hence the label.

The `pullback:` label indicates that the `@differentiating` function is a reverse-mode derivative function, or VJP. It returns a pullback function.

Currently, only reverse-mode differentiation is officially supported; forward-mode differentiation is off-by-default because it is not yet at feature parity with reverse-mode differentiation. Thus, this PR should use the `@differentiating` attribute to register pullback-returning reverse-mode derivatives. The compiler expects to find reverse-mode derivatives and will generate them if they do not exist, hence the CI failure (TF-429).

---

Sorry for the lack of documentation!

Eventually, when `@differentiable(linear)` functions and transposition are fully implemented, `@differentiating` attribute may be changed to only register differential-returning derivative functions.

The [differentiable programming manifesto](https://github.com/apple/swift/blob/master/docs/DifferentiableProgramming.md) describes the ideal final design for differentiable programming, so it doesn't mention the `pullback:` label, which may be removed.

Our [custom differentiation tutorial](https://github.com/tensorflow/swift/blob/master/docs/site/tutorials/custom_differentiation.ipynb) shows differentiation examples that work today, including derivative registration with the `pullback:` label, and more.

**Comment by vguerra** [comment]
Thank you @dan-zheng for the fix and explanation. Indeed it was not clear to me when to use the {{pullback}} label. I should have looked at the differentiation tutorial indeed. Thanks!

This way we should be able to register derivates for the trigonometric, hyperbolic and error functions using the {{@differentiating}} attribute as well. This reduces the amount of custom code in {{tgmath.swift.gyb}} file to achieve that. I am just making sure tests pass locally and will push a new commit addressing that.

**Comment by vguerra** [comment]
Oh .. I hit the problem of registering derivatives in a diff. file:

```console
/usr/local/ML/s4tf/compiler/swift/stdlib/public/Platform/tgmath.swift.gyb:576:2: error: derivative not in the same file as the original function
@differentiating(log1p)
~^~~~~~~~~~~~~~~~~~~~~~
```

So this is not possible ... yet :).

**Comment by dan-zheng** [comment]
> Oh .. I hit the problem of registering derivatives in a diff. file:
> 
> ```
> /usr/local/ML/s4tf/compiler/swift/stdlib/public/Platform/tgmath.swift.gyb:576:2: error: derivative not in the same file as the original function
> @differentiating(log1p)
> ~^~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> So this is not possible ... yet :).

Yes. This limitation will be lifted soon - there's ongoing work to support retroactive (including cross-module) derivative registration.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow macOS

**Review by dan-zheng — COMMENTED** [review]
@marcrasi: could you please help check whether the registered derivatives are ~mathematically correct? I think all of the newly differentiable functions have discontinuities (e.g. `floor`) but perhaps we can pick a sensible implementation or follow TensorFlow's precedent.

* `ceil`, `floor`, `round`, `trunc`
  * These are discontinuous at various points (integer values, or `x + 0.5` values for `round`).
  * TensorFlow defines these to return a zero gradient: [`CeilGrad`](https://github.com/tensorflow/tensorflow/blob/af019188adc704569b7cce51e50eba489af1a725/tensorflow/python/ops/math_grad.py#L1703), [`FloorGrad`](https://github.com/tensorflow/tensorflow/blob/af019188adc704569b7cce51e50eba489af1a725/tensorflow/python/ops/math_grad.py#L1698), [`RoundGrad`](https://github.com/tensorflow/tensorflow/blob/af019188adc704569b7cce51e50eba489af1a725/tensorflow/python/ops/math_grad.py#L1708), [`_TruncateDivGrad`](https://github.com/tensorflow/tensorflow/blob/af019188adc704569b7cce51e50eba489af1a725/tensorflow/python/ops/math_grad.py#L1313).
* `remainder`, `fmod` ([remainder from truncated division](https://developer.apple.com/documentation/swift/double/2885641-truncatingremainder))
  * I'm not sure `{ v in (v, v) }` is a correct pullback for these functions. [Math StackExchange results for "derivative of remainder"](https://math.stackexchange.com/search?q=derivative+of+remainder).

**Comment by vguerra** [comment]
Just a side note. I based the implementation of the derivatives according to what pytorch does [here](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml) : 

* [ceil](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L204)
* [floor](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L341)
* [round](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L729)
* [trunc](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L879)
* [remainder](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L710)
* [floor](https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L344)

but indeed I am not sure I understood the `reminder` and `floor` derivatives, will have a look at the links you sent .. and as well interested in what @marcrasi has to say :).

**Review by marcrasi — CHANGES_REQUESTED** [review]
I have no opinion about what to do at discontinuities because I don't know what the best practices for that are. If we're doing something that pytorch does, seems good to me.

File: test/stdlib/tgmath.swift.gyb [hunk 1]
```diff
@@ -273,6 +273,16 @@ MathTests.test("gradient_${T}") {
...
+  let remainderGrad = gradient(at: 4.0 as ${T}, 5.0 as ${T}, in: { x, y in remainder(x, y) })
+  expectEqualWithTolerance(1.0, remainderGrad.0, ulps: 16)
+  expectEqualWithTolerance(1.0, remainderGrad.1, ulps: 16)
+  let fmodGrad = gradient(at: 4.0 as ${T}, 5.0 as ${T}, in: { x, y in fmod(x, y) })
+  expectEqualWithTolerance(1.0, fmodGrad.0, ulps: 16)
+  expectEqualWithTolerance(1.0, fmodGrad.1, ulps: 16)
+  expectEqualWithTolerance(0.0, gradient(at: 2.0 as ${T}, in: { ceil($0) }), ulps: 16)
+  expectEqualWithTolerance(0.0, gradient(at: 2.0 as ${T}, in: { floor($0) }), ulps: 16)
+  expectEqualWithTolerance(0.0, gradient(at: 2.0 as ${T}, in: { round($0) }), ulps: 16)
+  expectEqualWithTolerance(0.0, gradient(at: 2.0 as ${T}, in: { trunc($0) }), ulps: 16)
```
[the remark this answers is not in the document]
Comment (marcrasi) [reply test/stdlib/tgmath.swift.gyb]: Actually half-integers are not good enough to avoid discontinuities. e.g. `0.5` divides `0.5`.

This loop body should work better:
```
guard y != 0 && abs(remainder(x, y)) > 0 else { continue }
checkGradient(remainder, x, y)
checkGradient(fmod, x, y)
```

**Comment by vguerra** [comment]
hi @marcrasi , thank you for the review & suggestions, i'll address them and push shortly a new commit.

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 2]
```diff
@@ -95,22 +84,75 @@ public func frexp<T: BinaryFloatingPoint>(_ x: T) -> (T, Int) {
...
   return (fma(x, y, z), { v in (v * y, v * x, v) })
 }
 
+@usableFromInline
+@differentiating(remainder)
+func _vjpRemainder<T: FloatingPoint & Differentiable> (
+  _ x: T,
+  _ y: T
+) -> (value: T, pullback: (T) -> (T, T)) where T == T.TangentVector {
+  return (remainder(x, y), { v in (v, v) })
```
[the remark this answers is not in the document]
Comment (vguerra) [reply stdlib/public/Platform/tgmath.swift.gyb]: Addressed on fc402e074639c6f42878817fe7b42f9a4f613bc5

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 3]
```diff
@@ -95,22 +84,75 @@ public func frexp<T: BinaryFloatingPoint>(_ x: T) -> (T, Int) {
...
+}
+
+@usableFromInline
+@differentiating(fmod)
+func _vjpFmod<T: FloatingPoint & Differentiable> (
+  _ x: T,
+  _ y: T
+) -> (value: T, pullback: (T) -> (T, T)) where T == T.TangentVector {
+  return (fmod(x, y), { v in (v, v) })
```
[the remark this answers is not in the document]
Comment (vguerra) [reply stdlib/public/Platform/tgmath.swift.gyb]: Addressed on fc402e074639c6f42878817fe7b42f9a4f613bc5

File: test/stdlib/tgmath.swift.gyb (hunk 1 above)
[the remark this answers is not in the document]
Comment (vguerra) [reply test/stdlib/tgmath.swift.gyb]: Addressed on fc402e074639c6f42878817fe7b42f9a4f613bc5

**Comment by vguerra** [comment]
I pushed fc402e074639c6f42878817fe7b42f9a4f613bc5 to address the issues with derivatives of `remainder` and `fmod`, plus the way we test them. All of this based on suggestions from @marcrasi ( thanks Marc! ).

I had to rebase the branch, sorry for that :/

File: test/stdlib/tgmath.swift.gyb [hunk 4]
```diff
@@ -273,6 +285,21 @@ MathTests.test("gradient_${T}") {
...
+  for a in -10...10 {
+    let x = ${T}(a)
+    for b in -10...10 {
+      let y = ${T}(b)
+      guard b != 0 && remainder(x, y).sign == remainder(x + ${T}(0.001), y).sign &&
+        remainder(x, y).sign == remainder(x, y + ${T}(0.001)).sign
```
Comment (vguerra) [inline test/stdlib/tgmath.swift.gyb]: I found out that for the combinations of `x` and `y` that don't satisfy this condition, computing the derivative for `remainder` using the definition of derivative would give unexpected results ( I guess due to the fact of it being discontinus ) hence I exclude them.

File: test/stdlib/tgmath.swift.gyb [hunk 5]
```diff
@@ -53,6 +53,18 @@ func expectEqualWithTolerance<T>(_ expected: TestLiteralType, _ actual: T,
              file: file, line: line)
 }
 
+func checkGradient<T: BinaryFloatingPoint & Differentiable>(
+  _ f: @differentiable (T, T) -> T,
+  _ x: T,
+  _ y: T) where T == T.TangentVector {
```
Comment (marcrasi) [inline test/stdlib/tgmath.swift.gyb]: style nit: `) where T == T.TangentVector {` or `{` should be moved to the next line, with indentation level 0, to give visual separation between the argument list and the function body

File: test/stdlib/tgmath.swift.gyb [hunk 6]
```diff
@@ -53,6 +53,18 @@ func expectEqualWithTolerance<T>(_ expected: TestLiteralType, _ actual: T,
...
+  let eps = T(0.001)
+  let grad = gradient(at: x, y, in: f)
+  let dfdx = (f(x + eps, y) - f(x, y)) / eps
+  let dfdy = (f(x, y + eps) - f(x, y)) / eps
+  expectEqualWithTolerance(TestLiteralType(dfdx.rounded(.toNearestOrAwayFromZero)), grad.0)
+  expectEqualWithTolerance(TestLiteralType(dfdy.rounded(.toNearestOrAwayFromZero)), grad.1)
```
Comment (marcrasi) [inline test/stdlib/tgmath.swift.gyb]: Is the rounding necessary? Can you make this work by making it more tolerant instead of by rounding?

More tolerance seems like a safer test than rounding, because rounding could mask for example a situation where the gradient is supposed to be `1.8` but we calculate it as `2`. (Based on my understanding, the gradients will actually always be integers, so this hypothetical situation will never actually happen. But tests are good for catching cases where we have misunderstood something, so it would be good to have a test that could catch such a situation.)

**Review by marcrasi — APPROVED** [review]

File: test/stdlib/tgmath.swift.gyb (hunk 5 above)
Comment (vguerra) [reply test/stdlib/tgmath.swift.gyb]: done in 7ce2e4c

File: test/stdlib/tgmath.swift.gyb (hunk 6 above)
Comment (vguerra) [reply test/stdlib/tgmath.swift.gyb]: Given the `eps` value I was using ( 0.001)  I needed to use tolerance of around 3700 for the check to pass, which seemed high to me, hence I opted for rounding, but you make a good point on having the ability to catch future errors here.

If I increase the eps to be `0.1` then the test pass with a tolerance of 32.

done in 7ce2e4c

**Comment by vguerra** [comment]
hi @dan-zheng , could you please trigger CI on this PR?

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by vguerra** [comment]
I am having a look at the tests that fail.

**Comment by vguerra** [comment]
`tgmath.swift.gyb`tests are now passing locally for me, could you please trigger CI once again? .. thanks!

**Comment by marcrasi** [comment]
@swift-ci Please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci Please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci Please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci Please test tensorflow

**Comment by vguerra** [comment]
checks passed :) ... would you agree that we could merge this now?

**Comment by marcrasi** [comment]
Yes, seems good to me!
